File: 208_Implement_Trie_Prefix/Trie.py
class TrieNode(object):
    def __init__(self):
        """
        Initialize your data structure here.
        :type children: dict{key:TrieNode}
        :typr key: str
        :type count: int (# word)
        """
        self.isKey = False
        self.children = {}

class Trie(object):
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        """
        Insert a word into a trie.
        :type word: str
        :rtype: void
        """
        prev_node = None
        node = self.root
        i = 0
        while i < len(word) and word[i] in node.children:
            prev_node = node
            node = node.children[word[i]]
            i += 1

        while i < len(word):
            node.children[word[i]] = TrieNode()
            prev_node = node
            node = node.children[word[i]]
            i += 1
        # True if this Node is the end of somestring
        prev_node.children[word[-1]].isKey = True


    def search(self, word):
        """
        Return if the word is in the trie.
        A word in a Trie must the the leaf of the tree
        :type word: str
        :rtype: bool
        """
        node = self.root
        i = 0
        while i < len(word) and word[i] in node.children:
            node = node.children[word[i]]
            i += 1
        if i == len(word) and node.isKey:
            return True
        else:
            return False

    def startsWith(self, prefix):
        """
        Return if there is any word in the trie
        that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        node = self.root
        i = 0
        while i < len(prefix) and prefix[i] in node.children:
            node = node.children[prefix[i]]
            i += 1
        return True if i == len(prefix) else False

trie = Trie()
trie.insert("abc")
print trie.search("abc")
print trie.search("ab")
trie.insert("ab")
print trie.search("ab")
trie.insert("ab")
print trie.search("ab")

File: 208_Implement_Trie_Prefix/Trie.cpp
#include <iostream>
#include <vector>
#define ALPHABET_SIZE 26


using namespace std;


class TrieNode {
	public:
    	// Initialize your data structure here.
    	TrieNode *next[ALPHABET_SIZE];
        bool isWord;
        TrieNode(bool b=false) {
            // void * memset ( void * ptr, int value, size_t num );
            // Sets the first num bytes of the block of memory pointed by ptr to the specified value
            memset(next, 0, sizeof(next));
            isWord = b;
    	}
};

class Trie {
	public:
    	Trie() {
        	root = new TrieNode();
    	}

    	// Inserts a word into the trie.
        void insert(string s) {
            TrieNode* p = root;
            for(int i = 0; i < s.size(); i++) {
                if(p -> next[s[i] - 'a'] == NULL)
                    p -> next[s[i] - 'a'] = new TrieNode();
                p = p -> next[s[i] - 'a'];
            }
            p -> isWord = true;
        }

        // Delete a key from a Trie
        void deleteKey(string key) {
            if(!key.empty()) {
                deleteHelper(this->root, key, 0, key.size());
            }
        }

        // Returns if the word is in the trie.
    	bool search(string key) {
            TrieNode* p = find(key);
            return p != NULL && p -> isWord;
    	}

    	// Returns if there is any word in the trie
    	// that starts with the given prefix.
    	bool startsWith(string prefix) {
            return find(prefix) != NULL;
    	}

        bool isLeafNode(TrieNode* p) {
            return p->isWord; 
        }

        // identify whether the node has next
        bool isFreeNode(TrieNode* p) {
            int i;
            for(i = 0; i < ALPHABET_SIZE; i++) {
                if(p -> next[i])
                    return false;
            }
            return true;
        }

	private:
        TrieNode* find(string key) {
            TrieNode* p = root;
            for(int i = 0; i < key.size() && p != NULL; i++)
                p = p -> next[key[i] - 'a'];
            return p;
        }
        
        // The helper function to delete a key from a Trie.
        // level: level (or depth) of a Trie
        bool deleteHelper(TrieNode *pNode, string key, int level, int len) {
            if(pNode) {
                // Base Case
                if(level == len) {
                    // Unmark leaf node
                    pNode -> isWord = false;

                    // If empty, node to be deleted
                    if(isFreeNode(pNode)) {
                        return true;
                    }
                    // not empty, should not delete the node.
                    return false;
                }
                else { // Recursive case
                    if(deleteHelper(pNode->next[key[level] - 'a'], key, level + 1, len)) {
                        // last node marked, delete it.
                        delete pNode->next[key[level] - 'a'];

                        // recursively climb up, and delete eligible node.
                        return ((!isLeafNode(pNode)) && isFreeNode(pNode));
                    }
                }
            }
            return false;
        }
    	TrieNode* root;
};

int main() {
    string keys[8] = {"she", "sells", "sea", "shore", "the", "by", "sheer"};
    Trie trie;
    for(int i = 0; i < 8; i++) {
        trie.insert(keys[i]);
    }
    cout << trie.search("she") << endl;
    trie.deleteKey(keys[0]);
    cout << trie.search("she") << endl;
    return 0;
}
